R/vol-surface-class.R

```diff
@@ -0,0 +1,173 @@
+#' VolSurface class
+#'
+#' The `VolSurface` class is designed to represent volatility surfaces.
+#' Checks whether the dimensions of the surface correspond to the right type of axis: moneyness, delta or strike
+#'
+#' @param reference_date `Date` value that represents the as of date.
+#' @param vol_quotes object of class `tibble` containing the volatility data.
+#' The tibble contains 3 columns: \cr
+#' 1. `maturity`: containing the maturities (tenors) of the vol surface, and is
+#' of class `character` with format "yyyymmdd", \cr
+#' 2. `strike` or `delta` or `moneyness`, this column is of class `numeric` and
+#' the name of this column informs on the type of the volatility surface. The moneyness
+#' is defined as strike/spot. \cr
+#' 3. `value`, this column contains the volatility values, and is of class
+#' `numeric`.
+#' @param ticker a label for the vol object, this should be similar ticker of the underlying index
+#' @param interpolation Interplation method, given as an object of class
+#' interpolation [fmbasics::Interpolation()].
+#' @export
+#' @seealso [interpolate.VolSurface]
+#' @examples vol_surface <- fmbasics::build_vol_surface()
+
+VolSurface <- function(reference_date, vol_quotes, ticker, interpolation) {
+  validate_VolSurface(new_VolSurface(
+    reference_date, vol_quotes, ticker, interpolation
+  ))
+}
+
+new_VolSurface <- function(reference_date, vol_quotes, ticker, interpolation) {
+  assertthat::assert_that(is.Interpolation(interpolation))
+
+  structure(
+    list(
+      reference_date = reference_date,
+      vol_quotes = vol_quotes,
+      ticker = ticker,
+      interpolation = interpolation
+    ),
+    class = "VolSurface"
+  )
+}
+
+
+validate_VolSurface <- function(x) {
+  assertthat::assert_that(
+    assertthat::is.date(x$reference_date),
+    colnames(x$vol_quotes)[1] == "maturity",
+    colnames(x$vol_quotes)[2] %in% c("strike", "delta", "moneyness"),
+    colnames(x$vol_quotes)[3] == "value",
+    tibble::is_tibble(x$vol_quotes),
+    is.character(x$vol_quotes$maturity),
+    assertthat::is.date(as.Date(x$vol_quotes$maturity, format = "%Y%m%d")),
+    is.double(x$vol_quotes$strike),
+    is.double(x$vol_quotes$value),
+    !is.unsorted(unique(x$vol_quotes$strike), na.rm = F),
+    !is.unsorted(as.Date(unique(x$vol_quotes$maturity), format = "%Y%m%d"), na.rm = F),
```

Comment:
Now these assertions are moved to VolQuotes object validation; and the user has to provide the data with sorted input dimensions similar to ZeroCurve, otherwise it would cause a big mess.